### PR TITLE
feat: Add workspace application support

### LINKED ISSUE #257: Workspace apps support
v2 workspaces need to support Dev URLs. For v2 community edition, a basic DevURL with private-only (local to the workspace and workspace owner) access is needed.

Advanced features with additional permissions and access levels, as well as wildcard support, will be part of paid Enterprise features. 

-----

- [x] https://github.com/coder/coder/issues/674
- [ ] https://github.com/coder/coder/issues/675
- [x] https://github.com/coder/coder/issues/676

### PR DESCRIPTION
**Pull request by kylecarbs**
This adds applications as a property to a workspace agent.

The resource is added to the Terraform provider here:
https://github.com/coder/terraform-provider-coder/pull/17

Apps will be opened in the dashboard or via the CLI
with `coder open <name>`. If `command` is specified, a
terminal will appear locally and on the web. If `target`
is specified, the browser will open to an exposed instance
of that target.

- [x] Fix odd context cancel bug on HTTP transport.

### CODE REVIEW

File: coderd/wsconncache/wsconncache.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+// Dialer creates a new agent connection by ID.
+type Dialer func(r *http.Request, id uuid.UUID) (*agent.Conn, error)
+
+// Conn wraps an agent connection with a reusable HTTP transport.
+type Conn struct {
+	*agent.Conn
+
+	locks         atomic.Uint64
```
Comment (spikecurtis) [inline coderd/wsconncache/wsconncache.go:40]: you increment and decrement this number, but never read it or use it for any conditional logic.

**Comment by spikecurtis** [comment]
At a high level, I also find it a bit strange that we're having coderd negotiate a WebRTC connection to the agent, when the agent is _already_ connected to coderd via the channel that the ICE negotiation is happening over.

File: coderd/wsconncache/wsconncache.go (hunk 1 above)
Comment (kylecarbs) [reply coderd/wsconncache/wsconncache.go:40]: Yup, this is all WIP.

File: coderd/wsconncache/wsconncache.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			}
+			c.connMutex.Lock()
+			delete(c.conns, id)
+			c.connMutex.Unlock()
+			// This should close after the delete so callers
+			// can check the `Closed()` channel for this to be expired.
+			_ = conn.CloseWithError(xerrors.New("cache timeout"))
+		}()
+		c.connMutex.Lock()
+		c.conns[id] = conn
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply coderd/wsconncache/wsconncache.go:103]: This is a good catch. Will certainly fix!

File: coderd/coderd.go [hunk 3]
```diff
@@ -93,6 +97,15 @@ func New(options *Options) *API {
 		tracing.HTTPMW(api.TracerProvider, "coderd.http"),
 	)
 
+	r.Route("/{user}/{workspaceagent}/{application}", func(r chi.Router) {
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply coderd/coderd.go:100]: Applications aren't accessed via the `/api/v2` route. They'll be served as if they're on the website, but really aren't.

eg. `/@kylecarbs/dev/apps/code-server`

File: provisionersdk/proto/provisioner.proto [hunk 4]
```diff
@@ -80,12 +80,22 @@ message Agent {
...
+        string token = 9;
+        string instance_id = 10;
     }
 }
 
+// App represents a dev-accessible application on the workspace.
+message App {
+    string name = 1;
+    string command = 2;
+    string url = 3;
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply provisionersdk/proto/provisioner.proto:94]: Good catch.

File: coderd/coderd.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (spikecurtis) [reply coderd/coderd.go:100]: I can see that.  My question is, why?  In particular, if they aren't on a versioned path, we may be painting our future selves into a corner

[the remark this answers is not in the document]
Comment (kylecarbs) [reply coderd/coderd.go:100]: I don't think we should expose users to API routes. I can't think of any applications that redirect a client to an API URL to serve content.

This isn't really an API endpoint either, it's much more of an HTTP proxy.

**Comment by spikecurtis** [comment]
> In a high availability deployment where multiple Coder replicas exist, there's no guarantee that the agent is connected to the dialed replica.

Ah, sure, but we have to establish a channel to exchange control messages.  So, before we can even start negotiating the WebRTC, we have a channel to the agent up.  Do we not want to use that channel in the hopes that WebRTC can find a path with fewer hops?

File: provisionersdk/proto/provisioner.proto [hunk 5]
```diff
@@ -80,12 +80,22 @@ message Agent {
     string operating_system = 5;
     string architecture = 6;
     string directory = 7;
+    repeated App apps = 8;
     oneof auth {
-        string token = 8;
-        string instance_id = 9;
+        string token = 9;
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply provisionersdk/proto/provisioner.proto:85]: It's nicer structurally from my perspective.

**Comment by kylecarbs** [comment]
We already do use that channel as the signal exchange for WebRTC. If we didn't, we could never establish a connection.

File: coderd/coderd.go [hunk 6]
```diff
@@ -93,6 +96,20 @@ func New(options *Options) *API {
...
+			httpmw.RateLimitPerMinute(options.APIRateLimit),
+			apiKeyMiddleware,
+			httpmw.ExtractUserParam(api.Database),
+		)
+		r.Get("/*", api.workspaceAppsProxyPath)
+	}
+	// %40 is the encoded character of the @ symbol. VS Code Web does
+	// not handle character encoding properly, so it's safe to assume
+	// other applications might not as well.
+	r.Route("/%40{user}/{workspacename}/apps/{workspaceapp}", apps)
```
Comment (vapurrmaid) [inline coderd/coderd.go:110]: Does this mean the FE should use `encodeURIComponent` ?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

File: coderd/coderd.go [hunk 7]
```diff
@@ -93,6 +96,20 @@ func New(options *Options) *API {
...
+	r.Route("/@{user}/{workspacename}/apps/{workspaceapp}", apps)
```
Comment (vapurrmaid) [inline coderd/coderd.go:111]: **Question**: Why do we have both then?

File: site/site.go [hunk 8]
```diff
@@ -17,6 +19,15 @@ import (
...
 )
 
+type apiResponseContextKey struct{}
+
+// WithAPIResponse returns a context with the APIResponse value attached.
+// This is used to inject API response data to the index.html for additional
+// metadata in error pages.
+func WithAPIResponse(ctx context.Context, apiResponse APIResponse) context.Context {
+	return context.WithValue(ctx, apiResponseContextKey{}, apiResponse)
+}
```
Comment (presleyp) [inline site/site.go:29]: Is this dev url specific or is it related to the error message discussion from yesterday?

File: coderd/coderd.go (hunk 6 above)
Comment (kylecarbs) [reply coderd/coderd.go:110]: Nah, we should always use `@` if possible. This is an edge-case for where certain client-side applications don't handle the `@` properly.

**Review by greyscaled — COMMENTED** [review]
FE = ✔️

File: coderd/rbac/README.md [hunk 9]
```diff
@@ -51,7 +51,7 @@ This can be represented by the following truth table, where Y represents *positi
 
 ## Example Permissions
 
-- `+site.*.*.read`: allowed to perform the `read` action against all objects of type `devurl` in a given Coder deployment.
```
Comment (f0ssel) [inline coderd/rbac/README.md:54]: kill it with 🔥

File: site/src/api/typesGenerated.ts [hunk 10]
```diff
@@ -420,6 +421,15 @@ export interface WorkspaceAgentResourceMetadata {
   readonly cpu_mhz: number
 }
 
+// From codersdk/workspaceapps.go:7:6
+export interface WorkspaceApp {
+  readonly id: string
+  readonly name: string
+  readonly command?: string
+  readonly access_url?: string
+  readonly icon: string
```
Comment (f0ssel) [inline site/src/api/typesGenerated.ts:430]: Is icon required? I think we should have some fallback on the frontend and make it optional in terraform if it isn't already.

**Review by f0ssel — APPROVED** [review]
Generally everything looks good to me. A little long so I didn't dive in super hard but the tests seem to cover enough bases to give me confidence.

File: site/src/api/typesGenerated.ts (hunk 10 above)
Comment (kylecarbs) [reply site/src/api/typesGenerated.ts:429]: True. I shall fix

**Pull request merged**